Library/Homebrew/livecheck/strategy.rb

```diff
@@ -80,6 +80,36 @@ def from_url(url, livecheck_strategy: nil, regex_provided: nil)
           (strategy.const_defined?(:PRIORITY) ? -strategy::PRIORITY : -DEFAULT_PRIORITY)
         end
       end
+
+      def self.page_headers(url)
```

Comment:
Unless we specifically need to use `curl` for retrieving headers, `#page_header` and `#page_contents` can seemingly share one network request, since they're both making `GET` requests to the same URL. As it stands, we're making a minimum of two requests to the same URL and it would multiply further if redirections are involved.

We could seemingly do something like this in both of these methods:

```ruby
URI.parse(url).open do |f|
  @headers = f.meta
  @page_contents = f.read
end
```

In this case, we wouldn't be making duplicate `GET` requests to the same URL when `#page_headers` and `#page_contents` are both used for a given strategy. Calling either of these would store both `@headers` and `@page_contents` and prevent subsequent network requests from being made for the same URL (for a given formula/cask).

One other nice aspect of this is that `OpenURI::Meta` is a hash that's easy to work with and we don't have to worry about parsing the headers:

```ruby
{
  "accept-ranges" => "bytes",
  "age" => "433349",
  "cache-control" => "max-age=604800",
  "content-type" => "text/html; charset=UTF-8",
  "date" => "Sun, 13 Dec 2020 23:36:51 GMT",
  "etag" => "\"3147526947\"",
  "expires" => "Sun, 20 Dec 2020 23:36:51 GMT",
  "last-modified" => "Thu, 17 Oct 2019 07:18:26 GMT",
  "server" => "ECS (dcb/7EEE)",
  "vary" => "Accept-Encoding",
  "x-cache" => "HIT",
  "content-length" => "648"
}
```

To be clear, `@headers` would be the response headers of the last request after any redirections are followed. If you need to get the `location` header before following the redirection, it's necessary to use `.open(redirect: false)` and handle the redirection manually like I'm doing in #9535. I may have overlooked something but it seems like this approach could be adapted to cover all of our needs without needing to duplicate requests.

Reply:
> To be clear, `@headers` would be the headers of the final response,

This will not work. I have already encountered some cases where the first header is a `302` redirect containing the needed `location` header, which is not contained in the final header which only contains e.g. the file name in `content-disposition`.

So currently, all headers are successively merged into one hash containing all headers.